06x-new-kafka-roller.md

```diff
@@ -0,0 +1,331 @@
+- controller-1
+- controller-2
+- combined-3
+- combined-4
+- combined-5
+- broker-6
+- broker-7
+- broker-8
+- broker-9
+- broker-10
+- broker-11
+
+1. The roller observes nodes and update their contexts based on the observation outcome:
+
+All the nodes except `combined-3` have the following Context with `nodeRef` being their `podname/node-id`, and `nodeRoles` having either `controller`, `broker` or both.
+```
+         nodeRef: controller-0/0
+         nodeRoles: controller
+         state: SERVING
+         lastTransition: 0123456
+         reason: MANUAL_ROLLING_UPDATE
+         numRestartAttempts: 0
+         numReconfigAttempts: 0
+         numRetries: 0
+```
+The `combined-3` node has the following context because it's ready from Kubernetes and KafkaAgent perspective but the operator could not establish an admin connection to it:
+```
+         nodeRef: combined-3/3
+         nodeRoles: controller,broker
+         state: NOT_RUNNING
+         lastTransition: 0123456
+         reason: POD_UNRESPONSIVE
+         numRestartAttempts: 0
+         numReconfigAttempts: 0
+         numRetries: 0
+```
+2. The roller checks if all of the controller nodes are combined and in `NOT_RUNNING` state. Since they are not and it has `POD_UNRESPONSIVE` reason, it restarts `combined-3` node and waits for it to have `SERVING` state. The `combined-3`'s context becomes:
+```
+         nodeRef: combined-3/3
+         nodeRoles: controller,broker
+         state: RESTARTED
+         lastTransition: 654987
+         reason: POD_UNRESPONSIVE
+         numRestartAttempts: 1
+         numReconfigAttempts: 0
+         numRetries: 0
+```
+3. `combined-3` state becomes `SERVING` and since its `numRestartAttempts` is greater than 1, the roller checks the rest of the nodes. 
+4. The roller checks which node is the active controller and finds that `controller-0` is. It then sends a request to the active controller via AdminClient to describe its `controller.quorum.fetch.timeout` config value.
+5. It then considers restarting `controller-1` and checks if the quorum health would be impacted. The operator sends a request to to the active controller to describe the quorum replication state. It finds that majority of the follower controllers have caught up with the quorum leader within the `controller.quorum.fetch.timeout.ms`. 
+6. The roller restarts `controller-1` as it has no impact on the quorum health. When it has `SERVING` state, the roller repeats the quorum check and restarts `controller-2` and then `controller-0`. 
+7. It then considers restarting `combined-4`, so it performs quorum healthcheck and then availability check. Both check passes therefore `combined-4` is restarted. The same is repeated for `combined-5`.
+8. All controller and combined nodes have `SERVING` state and `numRestartAttempts` set to 1, therefore the roller checks the broker nodes.
+9. It sends a request to describe all the topic partitions, and the following list of topics is returned:
```

Comment:
I think it will also need to query the broker config to get default min.insync.replica.